src/ansi-c/c_typecheck_expr.cpp

```diff
@@ -966,18 +973,20 @@ void c_typecheck_baset::typecheck_expr_sizeof(exprt &expr)
   {
     // This is a gcc extension.
     // https://gcc.gnu.org/onlinedocs/gcc-4.8.0/gcc/Pointer-Arith.html
-    new_expr = size_of_expr(char_type(), *this);
+    new_expr = from_integer(1, size_type());
   }
   else
   {
-    new_expr = size_of_expr(type, *this);
+    auto size_of_opt = size_of_expr(type, *this);
 
-    if(new_expr.is_nil())
+    if(!size_of_opt.has_value())
     {
       error().source_location = expr.source_location();
       error() << "type has no size: " << to_string(type) << eom;
       throw 0;
```

Comment:
This looks like it should be an invariant, judging from the error message.

Reply:
It is, but it's something that can be triggered by user input:
````c
struct S;
int main()
{
  return sizeof(S);
}
````
which also is the reason why we cannot make `size_of_expr` fail invariants.